packages/dds/tree/docs/indexes and branches.md

```diff
@@ -0,0 +1,142 @@
+
+-   On long lived branches.
+
+    Just like the main branch, if long lived branches are supported, they may needed indexes.
+    They may even need indexes for their corresponding working copies and remote branches if they are used collaboratively.
+
+## Optimizations Options
+
+Maintaining full in-memory versions of all indexes for all these locations all the time would be very expensive performance wise.
+
+This can be mitigated in several different ways:
+
+-   Maintain indexes only at places in the revision graph that require them.
+
+-   Use indexes when creating edits (tip of the "working copy" branch) but not when peer rebasing them.
+    For indexes using this approach this prevents remote edits from requiring indexes to be rebased.
+    This means that synchronously rebasing remote edits does not require index access, so index access can be async.
+    This also (when combined with he above) avoids maintaining most of the possible versions of an index.
+
+-   "Virtualized" indexes which lazily load data from blobs as needed, keeping only a portion of the data in memory at a given time.
+    This can make some accesses async, and thus works well with the above.
+
+-   Delta indexes: implements indexes as deltas to another version of the index which can be uses similarly to a write back cache or the tiering in level db.
+    This is a nice design pattern to handle the persisted vs in memory combination, but also generalizes.
+    This can be used to keep multiple in memory versions for nearby locations in the revision graph (ex: persisted, sequenced, local and per transaction).
+
+-   Copy on Write: Copy on write can also be used to optimize maintaining multiple indexes by sharing the parts that are the same.
+
+Note that this list is not compleat, but is sufficient to motivate the decisions below.
+
+## A General Approach
+
+Depending on the specific index, it might be needed at different places, and thus have very different usage patterns and optimization requirements.
+
+To keep as many options open as possible, shared tree's index abstraction can expose both the notion of branches and revisions to the index abstraction.
+This complicates the index abstraction a bit, but leaves the index implementation to pick if mutation or copy on write is a good approach, and if it should do virtualization.
+We then introduce the concept of an "IndexView", which accesses the index's information about a specific revision or branch.
+
+To make implementations of indexes simpler, helper code could be authored for both copy on write and mutation based index approaches (likely involving deltas as well) so this logic does not need to be implemented multiple times.
+
+To avoid having multiple data-structures tracking how different branches relate to each-other the we can generalize `Rebaser` into a `RevisionManager` can handle this once, and indexes can use it to track/notify/lookup anything needed related to that.
```

Comment:
```suggestion
To avoid having multiple data-structures tracking how different branches relate to each-other the we can generalize `Rebaser` into a `RevisionManager` that can handle this once, and indexes can use it to track/notify/lookup anything needed related to that.
```

Reply:
```suggestion
To avoid having multiple data-structures tracking how different branches relate to each-other, we can generalize `Rebaser` into a `RevisionManager` that can handle this once, and indexes can use it to track/notify/lookup anything needed related to that.
```